File: oc/AddressBookDemo/AddressBookDemo/controller/ContactsViewController.m
//
//  ContactsViewController.m
//  AddressBookDemo
//
//  Created by dfang on 2018-1-17.
//  Copyright © 2018年 east. All rights reserved.
//

#import "ContactsViewController.h"
#import "Contact.h"
#import "AddContactViewController.h"
#import "EditContactViewController.h"

@interface ContactsViewController () <AddContactViewControllerDelegate, EditContactViewControllerDelegate, UITableViewDelegate>

@property (weak, nonatomic) IBOutlet UITableView *tableView;
@property (nonatomic, copy) NSString *contactPath;
@property (nonatomic, strong) NSMutableArray *contacts;

@end

@implementation ContactsViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.
    self.title = [NSString stringWithFormat:@"%@的个人通讯录", self.name];
}

- (NSString *)contactPath {
    if (_contactPath == nil) {
        _contactPath = [[NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES) lastObject] stringByAppendingPathComponent:@"contacts.archiver"];
        NSLog(@"%@", _contactPath);
        // /var/mobile/Containers/Data/Application/C8899330-0926-4148-9CC0-9C8E181DF16A/Documents/contacts.archiver
        // /var/mobile/Containers/Data/Application/8BC1CA08-F459-4BF4-8545-134DE1DC0CC2/Documents/contacts.archiver
        // /var/mobile/Containers/Data/Application/D0F09697-B37F-4459-B71A-60B83BA969CC/Documents
    }
    return _contactPath;
}

- (NSMutableArray *)contacts {
    if (_contacts == nil) {
        _contacts = [NSKeyedUnarchiver unarchiveObjectWithFile:self.contactPath];
        if (_contacts == nil) {
            _contacts = [NSMutableArray array];
        }
    }
    return _contacts;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return self.contacts.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    // contactCellReuseId storyboard中设置
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"contactCellReuseId"];
    Contact *contact = self.contacts[indexPath.row];
    cell.textLabel.text = contact.name;
    cell.detailTextLabel.text = contact.phone;
    
    return cell;
}

- (UITableViewCellEditingStyle)tableView:(UITableView *)tableView editingStyleForRowAtIndexPath:(NSIndexPath *)indexPath {
    return UITableViewCellEditingStyleDelete;
}

- (void)tableView:(UITableView *)tableView commitEditingStyle:(UITableViewCellEditingStyle)editingStyle forRowAtIndexPath:(NSIndexPath *)indexPath {
    
    if (editingStyle == UITableViewCellEditingStyleDelete) {
        [self.contacts removeObjectAtIndex:indexPath.row];
        [self.tableView deleteRowsAtIndexPaths:@[indexPath] withRowAnimation:UITableViewRowAnimationFade];
        [NSKeyedArchiver archiveRootObject:self.contacts toFile:self.contactPath];
    }
}

- (IBAction)logout:(id)sender {
    [self.navigationController popViewControllerAnimated:YES];
}

- (IBAction)deleteItemClicked:(id)sender {
    [self.tableView setEditing:!self.tableView.isEditing animated:YES];
}

- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    id destVC = segue.destinationViewController;
    if ([destVC isKindOfClass:[AddContactViewController class]]) {
        AddContactViewController *addContactVC = destVC;
        addContactVC.delegate = self;
    } else if ([destVC isKindOfClass:[EditContactViewController class]]) {
        EditContactViewController *editContactVC = destVC;
        editContactVC.delegate = self;
        
        NSIndexPath *selectedIndex = self.tableView.indexPathForSelectedRow;
        Contact *contact = self.contacts[selectedIndex.row];
        editContactVC.contact = contact;
    }
}

- (void)addContactViewControllerDidSaveContact:(Contact *)contact {
    [self.contacts addObject:contact];
    NSIndexPath *indexPath = [NSIndexPath indexPathForRow:self.contacts.count - 1 inSection:0];
    [self.tableView insertRowsAtIndexPaths:@[indexPath] withRowAnimation:UITableViewRowAnimationFade];
    
    [self.navigationController popViewControllerAnimated:YES];
    [NSKeyedArchiver archiveRootObject:self.contacts toFile:self.contactPath];
}

- (void)editContactViewControllerDidSaveContact:(Contact *)contact {
    NSIndexPath *refreshPath = [NSIndexPath indexPathForRow: [self.contacts indexOfObject:contact] inSection:0];
    [self.tableView reloadRowsAtIndexPaths:@[refreshPath] withRowAnimation:UITableViewRowAnimationFade];
    
    [self.navigationController popViewControllerAnimated:YES];
    [NSKeyedArchiver archiveRootObject:self.contacts toFile:self.contactPath];
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

@end

File: swift/AddressBookDemo/AddressBookDemo/ContactsViewController.swift
//
//  ContactsViewController.swift
//  AddressBookDemo
//
//  Created by dfang on 2018-1-18.
//  Copyright © 2018年 east. All rights reserved.
//

import UIKit

class ContactsViewController: UITableViewController, AddContactViewControllerDelegate, EditContactViewControllerDelegate {
    
    @IBOutlet var tableview: UITableView!

    lazy var contactPath:String = {
        let docPath = NSSearchPathForDirectoriesInDomains(FileManager.SearchPathDirectory.documentDirectory, FileManager.SearchPathDomainMask.userDomainMask, true).last!
        return docPath + "/contacts.archiver"
    }()
    
    lazy var contacts: NSMutableArray = {
        var contacts = NSKeyedUnarchiver.unarchiveObject(withFile: contactPath) as? NSMutableArray
        if contacts == nil {
            contacts = NSMutableArray.init()
        }
        return contacts!
    }()
    
    var name:String?
    
    override func viewDidLoad() {
        self.title = name! + "的个人通讯录"
    }
    
    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return contacts.count
    }
    
    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: "contactCellReuseId")
        let contact:Contact = contacts.object(at: indexPath.row) as! Contact
        cell?.textLabel?.text = contact.name
        cell?.detailTextLabel?.text = contact.phone
        
        return cell!
    }
    
    override func tableView(_ tableView: UITableView, editingStyleForRowAt indexPath: IndexPath) -> UITableViewCellEditingStyle {
        return UITableViewCellEditingStyle.delete
    }
    
    override func tableView(_ tableView: UITableView, commit editingStyle: UITableViewCellEditingStyle, forRowAt indexPath: IndexPath) {
        if (editingStyle == UITableViewCellEditingStyle.delete) {
            contacts.removeObject(at: indexPath.row)
            tableView.deleteRows(at: [indexPath], with: UITableViewRowAnimation.fade)
            
            NSKeyedArchiver.archiveRootObject(contacts, toFile: contactPath)
        }
    }
    
    @IBAction func logoutItemClicked(_ sender: UIBarButtonItem) {
        navigationController?.popViewController(animated: true)
    }
    
    @IBAction func deleteItemClicked(_ sender: UIBarButtonItem) {
        tableview.setEditing(!tableview.isEditing, animated: true)
    }
    
    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        let vc = segue.destination
        if (vc.isKind(of: AddContactViewController.self)) {
            let addContactVC = vc as! AddContactViewController
            addContactVC.delegate = self
        } else if (vc.isKind(of: EditContactViewController.self)) {
            let editContactVC = vc as! EditContactViewController
            editContactVC.delegate = self
            
            let indexPath = tableview.indexPathForSelectedRow
            editContactVC.contact = contacts.object(at: (indexPath?.row)!) as! Contact
        }
    }
    
    // AddContactViewControllerDelegate
    func addContactViewControllerDidSaveContact(contact: Contact) {
        contacts.add(contact)
        let lastIndexPath = IndexPath.init(row: contacts.count - 1, section: 0)
        tableview.insertRows(at: [lastIndexPath], with: UITableViewRowAnimation.fade)
        navigationController?.popViewController(animated: true)
        
        NSKeyedArchiver.archiveRootObject(contacts, toFile: contactPath)
    }
    
    // EditContactViewControllerDelegate
    func editContactViewControllerDidSaveContact(contact: Contact) {
        let index = contacts.index(of: contact)
        let refreshIndexPath = IndexPath.init(row: index, section: 0)
        tableview.reloadRows(at: [refreshIndexPath], with: UITableViewRowAnimation.fade)
        navigationController?.popViewController(animated: true)
        
        NSKeyedArchiver.archiveRootObject(contacts, toFile: contactPath)
    }
    
}
